mate-calc/data/gcalctool.1
.\" Copyright (c) 1987-2008 - Sun Microsystems, Inc.
.TH GCALCTOOL 1 "17 June 2009"
.SH NAME
gcalctool \- a desktop calculator
.SH SYNOPSIS
.B gcalctool
[
.I OPTION
] ...
.SH DESCRIPTION
.B gcalctool
is the official calculator of the MATE desktop environment.
.SH OPTIONS
.TP
.B \-s, \-\-solve
Solve the equation provided following this option.
.TP
.B \-u, \-\-unittest
Run the unit tests.
.TP
.B \-v, \-\-version
Output version information and exit.
.TP
.B \-h, \-?, \-\-help
Prints the command line options.
.SH FILES
.PD 0
.TP 18
.B ~/.gcalctoolrc
user's personal gcalctool resources for customizing the appearance and color of
.B gcalctool
.sp
.LP


libmate/NEWS
libmate 2.32.0

mate-sound APIs now use libcanberra internally. That means that
mate_sound_connection_get() and will always return error (-1), and
mate_sound_sample_load also always return -1 but still cache
the sample using the libcanberra API.
libmate continues to link to libesound and libaudiofile for ABI
compatibility; use --disable-esd to remove this dependency.
(#603198, #603349)


libmate 2.30.0

* Enable silent build using AM_SILENT_RULES (Christian Persch)
* Put SlowKeys and BounceKeys default delays back to 300 (Willie Walker)
* Deprecate mate-sound (Christian Persch)
* Deprecate mate_[un]setenv (Christian Persch)
* Update the description of the picture_options key (Matthias Clasen)

* Updated translations
  - Khaled Hosny (ar)
  - astur (ast)
  - Alexander Shopov (bg)
  - Sadia Afroz (bn)
  - Jordi Mallach (ca)
  - Reşat SABIQ (crh)
  - Andre Klapper (cs)
  - Iestyn Pryce (cy)
  - Ask Hjorth Larsen (da)
  - Mario Blättermann (de)
  - Bakaoukas Nikolaos (el)
  - Thomas Thurman (en@shaw)
  - Bruce Cowan (en_GB)
  - Jorge González (es)
  - Mattias Põldaru (et)
  - Iñaki Larrañaga Murgoitio (eu)
  - Tommi Vainikainen (fi)
  - Claude Paroz (fr)
  - Fran Diéguez (gl)
  - Gabor Kelemen (hu)
  - Francesco Marletta (it)
  - Rimas Kudelis (lt)
  - Badral Sanligiin (mn)
  - Kjartan Maraas (nb)
  - Nils-Christoph Fiedler (nds)
  - Wouter Bolsterlee (nl)
  - A S Alam (pa)
  - Piotr Drąg (pl)
  - Duarte Loreto (pt)
  - Henrique P Machado (pt_BR)
  - Adi Roiban (ro)
  - Yuri Kozlov (ru)
  - Marcel Telka (sk)
  - Matej Urbančič (sl)
  - Горан Ракић (sr)
  - Goran Rakić (sr@latin)
  - Daniel Nylander (sv)
  - Maxim Dziumanenko (uk)
  - Clytie Siddall (vi)
  - Chao-Hsiung Liao (zh_HK)
  - Chao-Hsiung Liao (zh_TW)


libmate 2.28.0

Aron Xu (1):
      Corrected 5 strings.
Denis Arnaud (1):
      Updated breton translation and added br in LINGUAS
Lucian Adrian Grijincu (1):
      Updated Romanian translation
Miloš Popović (1):
      Updated Serbian translation
Rodrigo L. M. Flores (1):
      Updated Brazilian Portuguese mailing list address.

 
libmate 2.27.5

* Update default properties for menus_have_icons, buttons_have_icons and
  toolbar_style
* Port mate-sound.c to g_poll to fix portability for old OSX versions without
  poll.
* Update the address of the FSF by replacing old bits with new ones

* Updated translations
  - Jordi Mallach (ca)
  - Jordi Mallach (ca@valencia)
  - Khaled Hosny (ar)
  - Kjartan Maraas (nb)
  - Marcel Telka (sk)
  - Ray Wang (zh_CN)
  - Shankar Prasad (kn)
  - Sweta Kothari (gu)


libmate 2.26.0

* Updated translations
  (as, bn_IN, crh, cs, da, el, gl, hi, it, ja, lt, ml, mr,
   ro, ru, ta, te)

libmate 2.24.1

* Fix warnings about missing format specifiers (Kjartan)
* Remove #if 0'ed out code (Kjartan)
* Updated translations
  (ar, bg, en_GB, eu, fr, gu, lv, nl, or, pl, pt, th, tr,
   vi, zh_HK, zh_TW)

libmate 2.24.0.1

* Make thumbnail housekeeping less agressive by default
  (Michael Chudobiak)

libmate 2.24.0

* Updated translations
  (ar, da, gu, it, ro, ru, sr, ta, te)

libmate 2.23.92

* Port mate-open to GIO. (Jean-Yves Lefort)
* Build fix (Tor Lillqvist)

libmate 2.23.5

* Fixes for sound related issues (Jeff Steadfast)

libmate 2.23.4

* Updated translations.

libmate 2.23.3

* Code cleanups (Kjartan, Christian)
* Added schema to support the thumbnail cache cleaner (Michael J. Chudobiak)
* Plug leak (Stefan Kost)
* Add disable_application_handlers lockdown schema (William Jon McCann)
 
libmate 2.22.0

* Release 2.22.0
 
libmate 2.21.91

* Add schemas for tasks and calendar
* Use the GIO backend for the filechooser

libmate 2.21.90

* Set sound as enabled by default (Christian)
* Bypass a11y on win32 (Tor)
* Remove unused popt stuff (dmacks at netspace org) 
* Fix the toolbar icon size schema type (#401030)
* Win32 fixes
* SElinux fixes
* Warning fixes
* Translation updates

libmate 2.20.0

* Build fixes (Christian Persch)
* Set new background (Lucas Rocha, Vincent Untz)
* Fix module loading (Mariano Suárez-Alvarez, Christian Persch)
* Only depend on esd when esd is compiled in (Daniel Gryniewicz)
* Updated translations:
  - Albanian (Laurent Dhima)
  - Arabic (Khaled Hosny)
  - Assamese (Amitakhya Phukan)
  - Basque (Inaki Larranaga Murgoitio)
  - Bengali (Khandakar Mujahidul Islam)
  - Brazilian Portugese (Og Maciel)
  - Bulgarian (Alexander Shopov)
  - Canadian English (Adam Weinberger)
  - Catalan (Gil Forcada)
  - Danish (Kenneth Nielsen)
  - Estonian (Ivar Smolin)
  - Finnish (Ilkka Tuohela)
  - French (Stephane Raimbault)
  - Georgian (Vladimer Sichinava)
  - German (Hendrik Richter)
  - Gujarati (Ankit Patel)
  - Hungarian (Gabor Kelemen)
  - Italian (Francesco Marletta)
  - Japanese (Takeshi AIHANA)
  - Lithuanian (Žygimantas Beručka)
  - Macedonian (Jovan Naumovski)
  - Malayalam (Ani Peter)
  - Norwegian bokmål (Kjartan Maraas)
  - Polish (MATE Pl Team)
  - Portugese (Duarte Loreto)
  - Romaninan (Mişu Moldovan)
  - Russian (Nikolay V. Shmyrev)
  - Serbian (Goran Rakić)
  - Sinhala (Danishka Navin)
  - Spanish (Jorge Gonzalez)
  - Swedish (Daniel Nylander)
  - Thai (Theppitak Karoonboonyanan)
  - Ukrainian (Maxim Dziumanenko)
  - Vietnamese (Clytie Siddall)

libmate 2.19.1

* Don't initialise ALSA when sound is disabled (449249; Christian Persch)
* Translation updates: Runa Bhattacharjee (bn_IN), Hendrik Richter (de),
  Iñaki Larrañaga Murgoitio (eu), Ilkka Tuohela (fi), Ankit Patel (gu),
  Takeshi AIHANA (ja), Leonardo Fereira Fontenelle (pt_BR), Clytie Siddall (vi)

libmate 2.19.0

What's new since 2.18.0

* Rebuild the type builtins when the Makefile changes (Christian Persch)
* Make bug-buddy work with applets again (Matthias Clasen, #424949)
* Fix problems when building static libraries (Kjartan Maraas, #405352)
* Fix typos (Kenneth Nielsen, Kjartan Maraas, #429751)
* Add more information to docs for mate_program_init () (Christian Persch,
  #82943)
* Updated translations
  - mg: Thierry Randrianiriana
  - nb: Kjartan Maraas

mate-calc/src/gcalctool.c
/*  Copyright (c) 1987-2008 Sun Microsystems, Inc. All Rights Reserved.
 *  Copyright (c) 2008-2009 Robert Ancell
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 *  02111-1307, USA.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "currency.h"
#include "unittest.h"
#include "math-window.h"
#include "mp-equation.h"

static GSettings *settings = NULL;

static MathWindow *window;

static void
version(const gchar *progname)
{
    /* NOTE: Is not translated so can be easily parsed */
    fprintf(stderr, "%1$s %2$s\n", progname, VERSION);
}


static void
solve(const char *equation)
{
    MPEquationOptions options;
    MPErrorCode error;
    MPNumber result;
    char result_str[1024];

    memset(&options, 0, sizeof(options));
    options.base = 10;
    options.wordlen = 32;
    options.angle_units = MP_DEGREES;

    error = mp_equation_parse(equation, &options, &result, NULL);
    if(error == PARSER_ERR_MP) {
        fprintf(stderr, "Error: %s\n", mp_get_error());
        exit(1);
    }
    else if(error != 0) {
        fprintf(stderr, "Error: %s\n", mp_error_code_to_string(error));
        exit(1);
    }
    else {
        mp_cast_to_string(&result, 10, 10, 9, 1, result_str, 1024);
        printf("%s\n", result_str);
        exit(0);
    }
}


static void
usage(const gchar *progname, gboolean show_application, gboolean show_gtk)
{
    fprintf(stderr,
            /* Description on how to use gcalctool displayed on command-line */
            _("Usage:\n"
              "  %s — Perform mathematical calculations"), progname);

    fprintf(stderr,
            "\n\n");

    fprintf(stderr,
            /* Description on gcalctool command-line help options displayed on command-line */
            _("Help Options:\n"
              "  -v, --version                   Show release version\n"
              "  -h, -?, --help                  Show help options\n"
              "  --help-all                      Show all help options\n"
              "  --help-gtk                      Show GTK+ options"));
    fprintf(stderr,
            "\n\n");

    if (show_gtk) {
        fprintf(stderr,
                /* Description on gcalctool command-line GTK+ options displayed on command-line */
                _("GTK+ Options:\n"
                  "  --class=CLASS                   Program class as used by the window manager\n"
                  "  --name=NAME                     Program name as used by the window manager\n"
                  "  --screen=SCREEN                 X screen to use\n"
                  "  --sync                          Make X calls synchronous\n"
                  "  --gtk-module=MODULES            Load additional GTK+ modules\n"
                  "  --g-fatal-warnings              Make all warnings fatal"));
        fprintf(stderr,
                "\n\n");
    }

    if (show_application) {
        fprintf(stderr,
                /* Description on gcalctool application options displayed on command-line */
                _("Application Options:\n"
                  "  -u, --unittest                  Perform unit tests\n"
                  "  -s, --solve <equation>          Solve the given equation"));
        fprintf(stderr,
                "\n\n");
    }
}


static void
get_options(int argc, char *argv[])
{
    int i;
    char *progname, *arg;

    progname = g_path_get_basename(argv[0]);

    for (i = 1; i < argc; i++) {
        arg = argv[i];

        if (strcmp(arg, "-v") == 0 ||
            strcmp(arg, "--version") == 0) {
            version(progname);
            exit(0);
        }
        else if (strcmp(arg, "-h") == 0 ||
                 strcmp(arg, "-?") == 0 ||
                 strcmp(arg, "--help") == 0) {
            usage(progname, TRUE, FALSE);
            exit(0);
        }
        else if (strcmp(arg, "--help-all") == 0) {
            usage(progname, TRUE, TRUE);
            exit(0);
        }
        else if (strcmp(arg, "--help-gtk") == 0) {
            usage(progname, FALSE, TRUE);
            exit(0);
        }
        else if (strcmp(arg, "-s") == 0 ||
            strcmp(arg, "--solve") == 0) {
            i++;
            if (i >= argc) {
                fprintf(stderr,
                        /* Error printed to stderr when user uses --solve argument without an equation */
                        _("Argument --solve requires an equation to solve"));
                fprintf(stderr, "\n");
                exit(1);
            }
            else
                solve(argv[i]);
        }
        else if (strcmp(arg, "-u") == 0 ||
            strcmp(arg, "--unittest") == 0) {
            unittest();
        }
        else {
            fprintf(stderr,
                    /* Error printed to stderr when user provides an unknown command-line argument */
                    _("Unknown argument '%s'"), arg);
            fprintf(stderr, "\n");
            usage(progname, TRUE, FALSE);
            exit(1);
        }
    }
}


static void
quit_cb(MathWindow *window)
{
    MathEquation *equation;
    MathButtons *buttons;

    equation = math_window_get_equation(window);
    buttons = math_window_get_buttons(window);

    g_settings_set_int(settings, "accuracy", math_equation_get_accuracy(equation));
    g_settings_set_int(settings, "word-size", math_equation_get_word_size(equation));
    g_settings_set_int(settings, "base", math_buttons_get_programming_base(buttons));
    g_settings_set_boolean(settings, "show-thousands", math_equation_get_show_thousands_separators(equation));
    g_settings_set_boolean(settings, "show-zeroes", math_equation_get_show_trailing_zeroes(equation));
    g_settings_set_enum(settings, "number-format", math_equation_get_number_format(equation));
    g_settings_set_enum(settings, "angle-units", math_equation_get_angle_units(equation));
    g_settings_set_enum(settings, "button-mode", math_buttons_get_mode(buttons));
    g_settings_set_string(settings, "source-currency", math_equation_get_source_currency(equation));
    g_settings_set_string(settings, "target-currency", math_equation_get_target_currency(equation));
    g_settings_sync();

    currency_free_resources();
    gtk_main_quit();
}


int
main(int argc, char **argv)
{
    MathEquation *equation;
    int accuracy = 9, word_size = 64, base = 10;
    gboolean show_tsep = FALSE, show_zeroes = FALSE;
    DisplayFormat number_format;
    MPAngleUnit angle_units;
    ButtonMode button_mode;
    gchar *source_currency, *target_currency;

    g_type_init();

    bindtextdomain(GETTEXT_PACKAGE, LOCALE_DIR);
    bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
    textdomain(GETTEXT_PACKAGE);

    /* Seed random number generator. */
    srand48((long) time((time_t *) 0));

    get_options(argc, argv);

    settings = g_settings_new ("org.mate.gcalctool");
    accuracy = g_settings_get_int(settings, "accuracy");
    word_size = g_settings_get_int(settings, "word-size");
    base = g_settings_get_int(settings, "base");
    show_tsep = g_settings_get_boolean(settings, "show-thousands");
    show_zeroes = g_settings_get_boolean(settings, "show-zeroes");
    number_format = g_settings_get_enum(settings, "number-format");
    angle_units = g_settings_get_enum(settings, "angle-units");
    button_mode = g_settings_get_enum(settings, "button-mode");
    source_currency = g_settings_get_string(settings, "source-currency");
    target_currency = g_settings_get_string(settings, "target-currency");

    equation = math_equation_new();
    math_equation_set_accuracy(equation, accuracy);
    math_equation_set_word_size(equation, word_size);
    math_equation_set_show_thousands_separators(equation, show_tsep);
    math_equation_set_show_trailing_zeroes(equation, show_zeroes);
    math_equation_set_number_format(equation, number_format);
    math_equation_set_angle_units(equation, angle_units);
    math_equation_set_source_currency(equation, source_currency);
    math_equation_set_target_currency(equation, target_currency);
    g_free(source_currency);
    g_free(target_currency);

    gtk_init(&argc, &argv);

    window = math_window_new(equation);
    g_signal_connect(G_OBJECT(window), "quit", G_CALLBACK(quit_cb), NULL);
    math_buttons_set_programming_base(math_window_get_buttons(window), base);
    math_buttons_set_mode(math_window_get_buttons(window), button_mode); // FIXME: We load the basic buttons even if we immediately switch to the next type

    gtk_widget_show(GTK_WIDGET(window));
    gtk_main();

    return(0);
}
